publicodes/source/evaluation.tsx

```diff
@@ -38,10 +38,7 @@ export const mergeMissing = (left, right) =>
 	mergeWith(add, left || {}, right || {})
 
 export const evaluateNode = (cache, situation, parsedRules, node) => {
-	const evaluatedNode = node.evaluate
-		? node.evaluate(cache, situation, parsedRules, node)
-		: node
-	return evaluatedNode
+	return node.evaluate(cache, situation, parsedRules, node)
```

Comment:
L'étape d'après serait d'éviter de trimbaler ce quadruplet de paramètres à tous les étages de notre code d'évaluation.

Je ne suis pas certain de la bonne manière de le faire. Une possibilité c'est d'utiliser un modèle objet ou on évalue avec `node.evaluate()` (sans paramètre) et la fonction `evaluate` accède aux paramètres via `this` `this.situation`, `this.cache`, `this.node`. Qu'en pensez-vous ? On peut faire mieux ?

Reply:
J'ai l'intuition que l'on peut faire mieux en faisant une sorte d'inversion de contrôle où l'on ne ferait pas d'évaluation dans les fonctions `evaluate` de chaque mécanisme mais simplement une description de l'évaluation (aka un effet), et avoir une fonction `evaluate` globale qui prend la description de chaque évaluation et fournit les paramètres `cache`, `situation`, etc.

C'est probablement une grosse refacto. Mais cette approche me semble possible aussi, et probablement plus simple à mettre en œuvre, pour le parsing ; ce qui permettrait de supprimer la fonction `recurse`.

Reply:
`parsedRules`, `cache` et `situation` sont trois arguments qui font parti d'un contexte d'évaluation. Celui ci doit être passé de noeud en noeud lors de l'évaluation de l'AST, je ne vois pas comment faire autrement. En revanche, on peut imaginer une simplification pour le cas des mécanismes / expressions sans effet de bord.

Pour ceux-ci, la fonction evaluate n'a pas besoin du contexte. 

Par ailleurs, les noeuds de l'arbre syntaxique ne devrait pas porter la fonction d'evaluation, mais celle-ci pourrait être définie comme un décorateur de l'arbre. On aurait une fonction polymorphic evaluate avec différente implémentation en fonction du type du noeud du genre : 
```
evaluate(node: Rule, context) {
   applicable = evaluate(node.applicability, context)
   if (applicable) {
     evaluate(node.formule, context)
}
evaluate(node: ProductMecanism) {
   evaluate(node.assiette) && evaluate(node.assiette) * evaluate(node.taux)
}
evaluate(node:SommeMecanism) {
...
}
```


Par conséquent, je ne suis pas sûr qu'il faille ajouter une structure de representation intermediaire en plus de l'AST (ce que tu appelles effet). Nous avons déjà la description de l'évaluation en tant que tel. Si elle n'est pas suffisante, on peut imaginer des étapes de transformation supplémentaire directement sur l'AST pour transformer en une autre representation (par exemple transformer les mécanisme en opérations atomiques).

Reply:
En fait quand je parlais d'effet, je pensais à quelque chose comme les hooks de React pour pouvoir utiliser l'équivalent d'un `useContext(situation)` à n'importe quelle profondeur de l'évaluation sans avoir un passer explicitement le paramètre à tous les étages. Le terme officiel est "effet algébrique" et c'est supporté nativement par certains langages comme Eff https://www.eff-lang.org.

Mais en fait je crois qu'utiliser un objet pour garder les données de contexte `this.cache`, `this.situation` et `this.parsedRules` est la solution la plus logique en JavaScript. Je réfléchissais à des alternatives car on a toujours un peu en tête que c'est "mal" d'utiliser des `class`, mais elles ont certains cas d'usage appropriés comme un parseur ou un évaluateur dans lequel cela permet de garder des variables de contexte facilement accessibles dans toutes les méthodes (un peu comme des variables globales mais limitées à l'objet).

Pour que le `this` soit effectivement transmis à toutes les fonctions, il faut aussi remplacer l'appel `node.evaluate()` en `this.evaluate(node)`, ce qui nécessite d'avoir une fonction d'évaluation polymorphique comme tu l'expliques. L'implémentation la plus simple est d'avoir un nom pour chaque variante de la fonction `evaluate` et d'avoir un objet global qui indexe les noms et les fonctions correspondantes, le point d'entrée devenant :

```js
function evaluate(node) {
  return evaluationsFunctions[node.kind]()
}
```

La refacto est assez claire dans mon esprit et me semble bénéfique, je pense faire une PR prochainement.